Impossible to emit CommonJS modules when targeting ES6

It seems that the following line is [not necessary anymore with TypeScript 1.7](https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#support-for---target-es6-with---module):

https://github.com/TypeStrong/ts-loader/blob/ff15e76933e0f403e97f271f91577499add90f27/index.ts#L282

Removing it would allow us to target ES6 while still emitting CommonJS modules.

Out of curiousity I implemented this change locally and found it broke 4 tests:

```
  1) es6resolveParent should have the correct output:

      Uncaught AssertionError: bundle.js is different between actual and expected
      + expected - actual

       /***/ function(module, exports) {

        "use strict";

      + Object.defineProperty(exports, "__esModule", {
      +   value: true
      + });
      +
        function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

        var BaseComponent = function BaseComponent() {
          _classCallCheck(this, BaseComponent);

      at E:\source\ts-loader\test\run.js:234:28
      at Array.forEach (native)
      at Watching.handler (E:\source\ts-loader\test\run.js:223:39)
      at Watching._done (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:81:7)
      at Watching.<anonymous> (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:61:18)
      at Compiler.emitRecords (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:282:37)
      at Watching.<anonymous> (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:58:19)
      at E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:275:11
      at Compiler.applyPluginsAsync (E:\source\ts-loader\node_modules\webpack\node_modules\tapable\lib\Tapable.js:60:69)
      at Compiler.afterEmit (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:272:8)
      at Compiler.<anonymous> (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:267:14)
      at E:\source\ts-loader\node_modules\webpack\node_modules\async\lib\async.js:52:16
      at done (E:\source\ts-loader\node_modules\webpack\node_modules\async\lib\async.js:246:17)
      at E:\source\ts-loader\node_modules\webpack\node_modules\async\lib\async.js:44:16
      at E:\source\ts-loader\node_modules\webpack\node_modules\watchpack\node_modules\graceful-fs\graceful-fs.js:42:10
      at E:\source\ts-loader\node_modules\enhanced-resolve\node_modules\graceful-fs\graceful-fs.js:42:10
      at E:\source\ts-loader\node_modules\fs-extra\node_modules\graceful-fs\graceful-fs.js:42:10
      at E:\Source\ts-loader\node_modules\mocha\node_modules\glob\node_modules\graceful-fs\graceful-fs.js:104:5
      at FSReqWrap.oncomplete (fs.js:95:15)

  2) es6resolveParent should work with transpile:

      Uncaught AssertionError: bundle.js is different between actual and expected
      + expected - actual

       /***/ function(module, exports) {

        "use strict";

      + Object.defineProperty(exports, "__esModule", {
      +   value: true
      + });
      +
        function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

        var BaseComponent = function BaseComponent() {
          _classCallCheck(this, BaseComponent);

      at E:\source\ts-loader\test\run.js:234:28
      at Array.forEach (native)
      at Watching.handler (E:\source\ts-loader\test\run.js:223:39)
      at Watching._done (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:81:7)
      at Watching.<anonymous> (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:61:18)
      at Compiler.emitRecords (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:282:37)
      at Watching.<anonymous> (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:58:19)
      at E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:275:11
      at Compiler.applyPluginsAsync (E:\source\ts-loader\node_modules\webpack\node_modules\tapable\lib\Tapable.js:60:69)
      at Compiler.afterEmit (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:272:8)
      at Compiler.<anonymous> (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:267:14)
      at E:\source\ts-loader\node_modules\webpack\node_modules\async\lib\async.js:52:16
      at done (E:\source\ts-loader\node_modules\webpack\node_modules\async\lib\async.js:246:17)
      at E:\source\ts-loader\node_modules\webpack\node_modules\async\lib\async.js:44:16
      at E:\source\ts-loader\node_modules\webpack\node_modules\watchpack\node_modules\graceful-fs\graceful-fs.js:42:10
      at E:\source\ts-loader\node_modules\enhanced-resolve\node_modules\graceful-fs\graceful-fs.js:42:10
      at E:\source\ts-loader\node_modules\fs-extra\node_modules\graceful-fs\graceful-fs.js:42:10
      at E:\Source\ts-loader\node_modules\mocha\node_modules\glob\node_modules\graceful-fs\graceful-fs.js:104:5
      at FSReqWrap.oncomplete (fs.js:95:15)

  3) issue92 should have the correct output:

      Uncaught AssertionError: bundle.js is different between actual and expected
      + expected - actual

       /***/ function(module, exports, __webpack_require__) {

        'use strict';

      - var submodule_1 = __webpack_require__(1);
      - exports['default'] = submodule_1['default'];
      + Object.defineProperty(exports, '__esModule', {
      +   value: true
      + });

      + function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }
      +
      + var _submoduleSubmodule = __webpack_require__(1);
      +
      + var _submoduleSubmodule2 = _interopRequireDefault(_submoduleSubmodule);
      +
      + exports['default'] = _submoduleSubmodule2['default'];
      + module.exports = exports['default'];
      +
       /***/ },
       /* 1 */
       /***/ function(module, exports) {

        "use strict";

      + Object.defineProperty(exports, "__esModule", {
      +   value: true
      + });
        exports["default"] = "Hello from submodule";
      + module.exports = exports["default"];

       /***/ }
       /******/ ]);

      at E:\source\ts-loader\test\run.js:234:28
      at Array.forEach (native)
      at Watching.handler (E:\source\ts-loader\test\run.js:223:39)
      at Watching._done (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:81:7)
      at Watching.<anonymous> (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:61:18)
      at Compiler.emitRecords (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:282:37)
      at Watching.<anonymous> (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:58:19)
      at E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:275:11
      at Compiler.applyPluginsAsync (E:\source\ts-loader\node_modules\webpack\node_modules\tapable\lib\Tapable.js:60:69)
      at Compiler.afterEmit (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:272:8)
      at Compiler.<anonymous> (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:267:14)
      at E:\source\ts-loader\node_modules\webpack\node_modules\async\lib\async.js:52:16
      at done (E:\source\ts-loader\node_modules\webpack\node_modules\async\lib\async.js:246:17)
      at E:\source\ts-loader\node_modules\webpack\node_modules\async\lib\async.js:44:16
      at E:\source\ts-loader\node_modules\webpack\node_modules\watchpack\node_modules\graceful-fs\graceful-fs.js:42:10
      at E:\source\ts-loader\node_modules\enhanced-resolve\node_modules\graceful-fs\graceful-fs.js:42:10
      at E:\source\ts-loader\node_modules\fs-extra\node_modules\graceful-fs\graceful-fs.js:42:10
      at E:\Source\ts-loader\node_modules\mocha\node_modules\glob\node_modules\graceful-fs\graceful-fs.js:104:5
      at FSReqWrap.oncomplete (fs.js:95:15)

  4) issue92 should work with transpile:

      Uncaught AssertionError: bundle.js is different between actual and expected
      + expected - actual

       /***/ function(module, exports, __webpack_require__) {

        'use strict';

      - var submodule_1 = __webpack_require__(1);
      - exports['default'] = submodule_1['default'];
      + Object.defineProperty(exports, '__esModule', {
      +   value: true
      + });

      + function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }
      +
      + var _submoduleSubmodule = __webpack_require__(1);
      +
      + var _submoduleSubmodule2 = _interopRequireDefault(_submoduleSubmodule);
      +
      + exports['default'] = _submoduleSubmodule2['default'];
      + module.exports = exports['default'];
      +
       /***/ },
       /* 1 */
       /***/ function(module, exports) {

        "use strict";

      + Object.defineProperty(exports, "__esModule", {
      +   value: true
      + });
        exports["default"] = "Hello from submodule";
      + module.exports = exports["default"];

       /***/ }
       /******/ ]);

      at E:\source\ts-loader\test\run.js:234:28
      at Array.forEach (native)
      at Watching.handler (E:\source\ts-loader\test\run.js:223:39)
      at Watching._done (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:81:7)
      at Watching.<anonymous> (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:61:18)
      at Compiler.emitRecords (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:282:37)
      at Watching.<anonymous> (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:58:19)
      at E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:275:11
      at Compiler.applyPluginsAsync (E:\source\ts-loader\node_modules\webpack\node_modules\tapable\lib\Tapable.js:60:69)
      at Compiler.afterEmit (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:272:8)
      at Compiler.<anonymous> (E:\source\ts-loader\node_modules\webpack\lib\Compiler.js:267:14)
      at E:\source\ts-loader\node_modules\webpack\node_modules\async\lib\async.js:52:16
      at done (E:\source\ts-loader\node_modules\webpack\node_modules\async\lib\async.js:246:17)
      at E:\source\ts-loader\node_modules\webpack\node_modules\async\lib\async.js:44:16
      at E:\source\ts-loader\node_modules\webpack\node_modules\watchpack\node_modules\graceful-fs\graceful-fs.js:42:10
      at E:\source\ts-loader\node_modules\enhanced-resolve\node_modules\graceful-fs\graceful-fs.js:42:10
      at E:\source\ts-loader\node_modules\fs-extra\node_modules\graceful-fs\graceful-fs.js:42:10
      at E:\Source\ts-loader\node_modules\mocha\node_modules\glob\node_modules\graceful-fs\graceful-fs.js:104:5
      at FSReqWrap.oncomplete (fs.js:95:15)
```

I'll be honest; my knowledge of what Webpack should be emitting is pretty non-existent.  Does this sound right?  I'm happy to amend the tests to comply with the changed emit but I'm not sure it is in this case.

--- a/index.ts
+++ b/index.ts
@@ -197,7 +197,6 @@ function ensureTypeScriptInstance(loaderOptions: LoaderOptions, loader: any): { 
     };
 
     var compilerOptions: typescript.CompilerOptions = {
-        module: 1 /* CommonJS */
     };
 
     // Load any available tsconfig.json file
@@ -275,14 +274,14 @@ function ensureTypeScriptInstance(loaderOptions: LoaderOptions, loader: any): { 
     instance.compilerOptions = objectAssign<typescript.CompilerOptions>(compilerOptions, configParseResult.options);
     filesToLoad = configParseResult.fileNames;
 
-    var libFileName = 'lib.d.ts';
-
-    // Special handling for ES6 targets
-    if (compilerOptions.target == 2 /* ES6 */) {
-        compilerOptions.module = 0 /* None */;
-        libFileName = 'lib.es6.d.ts';
+    // if `module` is not specified and not using ES6 target, default to CJS module output
+    if (compilerOptions.module == null && compilerOptions.target !== 2 /* ES6 */) {
+        compilerOptions.module = 1 /* CommonJS */
     }
-    libFileName = path.join(path.dirname(require.resolve(loaderOptions.compiler)), libFileName);
+    // special handling for TS 1.6 and target: es6
+    else if (compilerCompatible && semver.lt(compiler.version, '1.7.3-0') && compilerOptions.target == 2 /* ES6 */) {
+        compilerOptions.module = 0 /* None */;
+    }
 
     if (loaderOptions.transpileOnly) {
         // quick return for transpiling
@@ -295,10 +294,6 @@ function ensureTypeScriptInstance(loaderOptions: LoaderOptions, loader: any): { 
             formatErrors(diagnostics, instance, {file: configFilePath || 'tsconfig.json'}));
 
         return { instance: instances[loaderOptions.instance] = { compiler, compilerOptions, loaderOptions, files }};
-    }
-
-    if (!compilerOptions.noLib) {
-        filesToLoad.push(libFileName);
     }
 
     // Load initial files (core lib files, any files specified in tsconfig.json)
@@ -356,7 +351,7 @@ function ensureTypeScriptInstance(loaderOptions: LoaderOptions, loader: any): { 
         },
         getCurrentDirectory: () => process.cwd(),
         getCompilationSettings: () => compilerOptions,
-        getDefaultLibFileName: options => libFileName,
+        getDefaultLibFileName: options => compiler.getDefaultLibFilePath(options),
         getNewLine: () => newLine,
         log: log,
         resolveModuleNames: (moduleNames: string[], containingFile: string) => {

--- a/test/es6withCJS/app.ts
+++ b/test/es6withCJS/app.ts
@@ -1,0 +1,1 @@
+export default 'a';

--- a/test/es6withCJS/expectedOutput-1.6/output.txt
+++ b/test/es6withCJS/expectedOutput-1.6/output.txt
@@ -1,0 +1,6 @@
+
+ERROR in ./.test/es6withCJS/app.ts
+Module parse failed: index.js!app.ts Line 1: Unexpected token
+You may need an appropriate loader to handle this file type.
+| export default 'a';
+| 

--- a/test/es6withCJS/expectedOutput-1.7/bundle.js
+++ b/test/es6withCJS/expectedOutput-1.7/bundle.js
@@ -1,0 +1,51 @@
+/******/ (function(modules) { // webpackBootstrap
+/******/ 	// The module cache
+/******/ 	var installedModules = {};
+
+/******/ 	// The require function
+/******/ 	function __webpack_require__(moduleId) {
+
+/******/ 		// Check if module is in cache
+/******/ 		if(installedModules[moduleId])
+/******/ 			return installedModules[moduleId].exports;
+
+/******/ 		// Create a new module (and put it into the cache)
+/******/ 		var module = installedModules[moduleId] = {
+/******/ 			exports: {},
+/******/ 			id: moduleId,
+/******/ 			loaded: false
+/******/ 		};
+
+/******/ 		// Execute the module function
+/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
+
+/******/ 		// Flag the module as loaded
+/******/ 		module.loaded = true;
+
+/******/ 		// Return the exports of the module
+/******/ 		return module.exports;
+/******/ 	}
+
+
+/******/ 	// expose the modules object (__webpack_modules__)
+/******/ 	__webpack_require__.m = modules;
+
+/******/ 	// expose the module cache
+/******/ 	__webpack_require__.c = installedModules;
+
+/******/ 	// __webpack_public_path__
+/******/ 	__webpack_require__.p = "";
+
+/******/ 	// Load entry module and return exports
+/******/ 	return __webpack_require__(0);
+/******/ })
+/************************************************************************/
+/******/ ([
+/* 0 */
+/***/ function(module, exports) {
+
+	exports.default = 'a';
+
+
+/***/ }
+/******/ ]);

--- a/test/es6withCJS/expectedOutput-1.7/output.txt
+++ b/test/es6withCJS/expectedOutput-1.7/output.txt
@@ -1,0 +1,4 @@
+    Asset     Size  Chunks             Chunk Names
+bundle.js  1.41 kB       0  [emitted]  main
+chunk    {0} bundle.js (main) 23 bytes [rendered]
+    [0] ./.test/es6withCJS/app.ts 23 bytes {0} [built]

--- a/test/es6withCJS/expectedOutput-1.8/bundle.js
+++ b/test/es6withCJS/expectedOutput-1.8/bundle.js
@@ -1,0 +1,53 @@
+/******/ (function(modules) { // webpackBootstrap
+/******/ 	// The module cache
+/******/ 	var installedModules = {};
+
+/******/ 	// The require function
+/******/ 	function __webpack_require__(moduleId) {
+
+/******/ 		// Check if module is in cache
+/******/ 		if(installedModules[moduleId])
+/******/ 			return installedModules[moduleId].exports;
+
+/******/ 		// Create a new module (and put it into the cache)
+/******/ 		var module = installedModules[moduleId] = {
+/******/ 			exports: {},
+/******/ 			id: moduleId,
+/******/ 			loaded: false
+/******/ 		};
+
+/******/ 		// Execute the module function
+/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
+
+/******/ 		// Flag the module as loaded
+/******/ 		module.loaded = true;
+
+/******/ 		// Return the exports of the module
+/******/ 		return module.exports;
+/******/ 	}
+
+
+/******/ 	// expose the modules object (__webpack_modules__)
+/******/ 	__webpack_require__.m = modules;
+
+/******/ 	// expose the module cache
+/******/ 	__webpack_require__.c = installedModules;
+
+/******/ 	// __webpack_public_path__
+/******/ 	__webpack_require__.p = "";
+
+/******/ 	// Load entry module and return exports
+/******/ 	return __webpack_require__(0);
+/******/ })
+/************************************************************************/
+/******/ ([
+/* 0 */
+/***/ function(module, exports) {
+
+	"use strict";
+	Object.defineProperty(exports, "__esModule", { value: true });
+	exports.default = 'a';
+
+
+/***/ }
+/******/ ]);

--- a/test/es6withCJS/expectedOutput-1.8/output.txt
+++ b/test/es6withCJS/expectedOutput-1.8/output.txt
@@ -1,0 +1,4 @@
+    Asset     Size  Chunks             Chunk Names
+bundle.js  1.49 kB       0  [emitted]  main
+chunk    {0} bundle.js (main) 100 bytes [rendered]
+    [0] ./.test/es6withCJS/app.ts 100 bytes {0} [built]

--- a/test/es6withCJS/tsconfig.json
+++ b/test/es6withCJS/tsconfig.json
@@ -1,0 +1,6 @@
+{
+    "compilerOptions": {
+        "target": "es6",
+        "module": "commonjs"
+    }
+}

--- a/test/es6withCJS/webpack.config.js
+++ b/test/es6withCJS/webpack.config.js
@@ -1,0 +1,17 @@
+module.exports = {
+    entry: './app.ts',
+    output: {
+        filename: 'bundle.js'
+    },
+    resolve: {
+        extensions: ['', '.ts', '.js']
+    },
+    module: {
+        loaders: [
+            { test: /\.ts$/, loader: 'ts-loader' }
+        ]
+    }
+}
+
+// for test harness purposes only, you would not need this in a normal project
+module.exports.resolveLoader = { alias: { 'ts-loader': require('path').join(__dirname, "../../index.js") } }

--- a/test/ignoreDiagnostics/expectedOutput-1.8/output.txt
+++ b/test/ignoreDiagnostics/expectedOutput-1.8/output.txt
@@ -4,5 +4,4 @@ chunk    {0} bundle.js (main) 247 bytes [rendered]
     [0] ./.test/ignoreDiagnostics/app.ts 247 bytes {0} [built] [1 error]
 
 ERROR in ./.test/ignoreDiagnostics/app.ts
-[37m([39m[36m9[39m,[36m5[39m): [31merror TS2322: Type 'string' is not assignable to type 'Number'.
-  Property 'toFixed' is missing in type 'String'.[39m
+[37m([39m[36m9[39m,[36m5[39m): [31merror TS2322: Type 'string' is not assignable to type 'Number'.[39m

